internal/llm/agent/mcp-tools.go

```diff
@@ -310,33 +330,71 @@ func doGetMCPTools(ctx context.Context, permissions permission.Service, cfg *con
 			if err != nil {
 				return
 			}
+
 			mcpClients.Set(name, c)
 
-			tools := getTools(ctx, name, permissions, c, cfg.WorkingDir())
-			updateMCPState(name, MCPStateConnected, nil, c, len(tools))
+			tools := getTools(ctx, name, c)
 			result.Append(tools...)
+			updateMcpTools(name, tools)
+			updateMCPState(name, MCPStateConnected, nil, c, len(tools))
 		}(name, m)
 	}
 	wg.Wait()
+
 	return slices.Collect(result.Seq())
 }
 
+// updateMcpTools updates the global mcpTools and mcpClientTools maps
+func updateMcpTools(mcpName string, tools []tools.BaseTool) {
+	toolNames := make([]string, 0, len(tools))
+	for _, tool := range tools {
+		name := tool.Name()
+		if _, ok := mcpTools.Get(name); !ok {
+			slog.Info("Added MCP tool", "name", name, "mcp", mcpName)
+		}
+		mcpTools.Set(name, tool)
+		toolNames = append(toolNames, name)
+	}
+
+	// remove the tools that are no longer available
+	old, ok := mcpClientTools.Get(mcpName)
+	if ok {
+		slices.Sort(toolNames)
+		for _, name := range old {
+			if _, ok := slices.BinarySearch(toolNames, name); !ok {
+				mcpTools.Del(name)
+				slog.Info("Removed MCP tool", "name", name, "mcp", mcpName)
+			}
+		}
+	}
```

Comment:
i wonder if it isn't simpler to clear the old tool map and just set the new tools into it